plugin/keyring_vault/vault_keyring.cc

```
@@ -0,0 +1,219 @@
+  "The path to the keyring_vault configuration file",          /* comment    */
+  check_keyring_file_data,                                     /* check()    */
+  update_keyring_file_data,                                    /* update()   */
+  ""                                                           /* default    */
+);
+
+static struct st_mysql_sys_var *keyring_vault_system_variables[]= {
+  MYSQL_SYSVAR(config),
+  NULL
+};
+
+static int keyring_vault_init(MYSQL_PLUGIN plugin_info)
+{
+  try
+  {
+#ifdef HAVE_PSI_INTERFACE
+    keyring_init_psi_keys();
+#endif
+    if (init_keyring_locks())
+      return TRUE;
+
+    logger.reset(new Logger(plugin_info));
+    keys.reset(new Vault_keys_container(logger.get()));
+    boost::movelib::unique_ptr<IVault_curl> vault_curl(new Vault_curl(logger.get()));
+    boost::movelib::unique_ptr<IVault_parser> vault_parser(new Vault_parser(logger.get()));
+    IKeyring_io *keyring_io= new Vault_io(logger.get(), vault_curl.release(),
+                                          vault_parser.release());
+    if (keys->init(keyring_io, keyring_vault_config_file))
+    {
+      is_keys_container_initialized = FALSE;
+      logger->log(MY_ERROR_LEVEL, "keyring_vault initialization failure. Please check that"
+        " the keyring_vault_config_file points to readable keyring_vault configuration"
+        " file. Please also make sure Vault is running and accessible."
+        " The keyring_vault will stay unusable until correct configuration file gets"
+        " provided.");
+      return FALSE;
+    }
+    is_keys_container_initialized = TRUE;
+    return FALSE;
+  }
+  catch (const std::bad_alloc &e)
+  {
+    DBUG_ASSERT(0);
+    if (logger != NULL)
+      logger->log(MY_ERROR_LEVEL, "keyring_vault initialization failure "
+                                  "due to memory allocation failure");
+    return TRUE;
+  }
+  catch (...)
+  {
+    DBUG_ASSERT(0);
```

Extract to some helper function handle_unknown_exception, consider re-throwing at the end

---

Extracted to a helper function. I do not re-throw the exception as it would go into server code - were exceptions are banned - could cause server to crash instead of returning failure.

---

Right, no rethrowing here